Feature/877 add battery capacity fault capability

Yume27 opened the pull request:
* **Tickets addressed:** bsk-877
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Squash and Merge  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? How are your
commits organized? -->
This added the capability to simulate a fault in the `simpleBattery` module that reduces the actual storage capacity without directly altering the stated capacity.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? If you didn't 
add or update any tests justify this choice. -->
Added unit test to verify that 
1. Check if the battery capacity matches its nominal value when a fault occurs.
2. Verify that the charged capacity does not exceed the faulted capacity or drop below zero.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and 
completeness? -->
Updated rst file associated with simpleBattery module.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
None

juan-g-bonilla commented on src/simulation/power/simpleBattery/simpleBattery.cpp:
```diff
@@ -59,6 +59,16 @@ void SimpleBattery::evaluateBatteryModel(PowerStorageStatusMsgPayload *msg) {
         this->storedCharge = 0;
     }
 
+    if(this->fault == true) {   
```
Is there some value to making these `fault` and `faultCapacityRatio` input messages instead of class variables?

schaubh replied:
Mm, you are thinking like a battery configuration message?  I'm not sure I can think of an associated scenario that would need this.  I'll think on this.

schaubh replied:
Mm, interesting.  I could see value in that.  @Yume27 , what you can do is move the fault state to a message.  Then, in `simpleBattery` you check if this optional message is connected.  If yes, you read in the values and use them?  In your event handler you would just change the value of this input message.

schaubh replied:
Also, couldn't this fault feature be added to the `powerStorageBase()` base class?  This way all batteries based on this parent class would get this capability?

juan-g-bonilla commented on src/simulation/power/simpleBattery/simpleBattery.cpp:
```diff
@@ -59,6 +59,16 @@ void SimpleBattery::evaluateBatteryModel(PowerStorageStatusMsgPayload *msg) {
     }
 
+    if(this->fault == true) {   
+        if(this->faultCapacityRatio < 0)
+        {
+            this->faultCapacityRatio = 1.0;
+        }
+        if(this->storedCharge > this->storageCapacity * this->faultCapacityRatio) {
+            this->storedCharge = this->storageCapacity * this->faultCapacityRatio;
+        }
```
Could this be simplified and made more strict regarding acceptable values of `faultCapacityRatio`?
```
if (this->faultCapacityRatio < 0 || this->faultCapacityRatio > 1)
    bskLogger.bskLog(BSK_ERROR, "faultCapacityRatio should be between 0 and 1!");
else
    this->storedCharge = this->storageCapacity * this->faultCapacityRatio;
```

schaubh replied:
I agree.  Right now this code silently changes the value if <0 and the user is not notified.  I like the more compact version that doesn't silently change values.

schaubh replied:
@juan-g-bonilla , your code actually always set the stored charge to the max value.  I think the code should be:

```
if (this->faultCapacityRatio < 0 || this->faultCapacityRatio > 1) {
    bskLogger.bskLog(BSK_ERROR, "faultCapacityRatio should be between 0 and 1!");
} else if(this->storedCharge > this->storageCapacity * this->faultCapacityRatio) {
            this->storedCharge = this->storageCapacity * this->faultCapacityRatio;
        }
}
```

schaubh commented on src/simulation/power/simpleBattery/simpleBattery.h:
```diff
@@ -39,6 +39,8 @@ class SimpleBattery: public PowerStorageBase {
 
 public:
     double storageCapacity; //!< [W-s] Battery capacity in Watt-seconds (Joules).
+    bool fault = false; //!< Fault flag
```
These variables should be private variables, and then we need associated setter and getter functions

juan-g-bonilla replied:
If you’re using setters, then i would move the check `0<=faultCapacityRatio<=1` to the setter of `faultCapacityRatio` instead of every time the model is run.

schaubh replied:
But, if we are reading in the value from a message, then we don't have that control. I like the idea of moving the fault info to a battery fault message.  I think it will provide a more elegant solution.

schaubh commented on src/simulation/power/simpleBattery/simpleBattery.h:
```diff
@@ -39,6 +39,8 @@ class SimpleBattery: public PowerStorageBase {
 public:
     double storageCapacity; //!< [W-s] Battery capacity in Watt-seconds (Joules).
+    bool fault = false; //!< Fault flag
+    double faultCapacityRatio = 1.0; //!< Fault capacity ratio
```
explain what 1.0 default value means here.  This means by default we have full capacity, correct?

schaubh commented on src/simulation/power/simpleBattery/_UnitTest/test_simpleBatteryFault.py:
```diff
@@ -0,0 +1,119 @@
+#
+#  ISC License
+#
+#  Copyright (c) 2016, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
```
update copyright year to 2024

schaubh commented on src/simulation/power/simpleBattery/_UnitTest/test_simpleBatteryFault.py:
```diff
@@ -0,0 +1,119 @@
+    test_battery = simpleBattery.SimpleBattery()
+    test_battery.storedCharge_Init = storedChargeInit
+    test_battery.storageCapacity = batteryCapacity  
+    test_battery.fault = True
```
Right now you are only testing the functionality if there is a fault.  You should include a trivial test that checks that the default behavior is no fault.

schaubh commented on src/simulation/power/simpleBattery/simpleBattery.rst:
```diff
@@ -36,3 +36,8 @@ The next step is to attach one or more :ref:`PowerNodeUsageMsgPayload` instances
 
 For more information on how to set up and use this module, see the simple power system example :ref:`scenarioPowerDemo`.
+
+To simulate a battery capacity fault that reduces the actual storage capacity (without directly altering the stated capacity), users must set up the fault flag::
```
This is not very clear what we are doing here.  Can you please be more explicit on what is done when a fault is modeled?

schaubh commented on src/simulation/power/simpleBattery/simpleBattery.rst:
```diff
@@ -36,3 +36,8 @@ The next step is to attach one or more :ref:`PowerNodeUsageMsgPayload` instances
+
+To simulate a battery capacity fault that reduces the actual storage capacity (without directly altering the stated capacity), users must set up the fault flag::
+
+   battery.fault = True
```
This should be updated to use the private class setter function.  Check out 
https://avslab.github.io/basilisk/Documentation/moduleTemplates/cppModuleTemplate/cppModuleTemplate.html
to see how to name and document the setter and getter functions.

schaubh commented:
@Yume27 , your pre-commit test failed.  Be sure to install the pre-commit support by reading https://avslab.github.io/basilisk/Support/Developer/CodingGuidlines.html.  Look near the bottom for the link to CONTRIBUTING.md.  It talks about how to include the checks, and how to many force check specific files if needed.

juan-g-bonilla commented on src/simulation/power/simpleBattery/simpleBattery.cpp:
```diff
@@ -59,6 +59,16 @@ void SimpleBattery::evaluateBatteryModel(PowerStorageStatusMsgPayload *msg) {
         this->storedCharge = 0;
     }
 
+    if(this->fault == true) {   
```
Actually, is there a reason to have both `fault` AND `faultCapacityRatio`? Doesn’t `faultCapacityRatio==1` imply no fault in the capacity and anything below 1 implies a fault?

juan-g-bonilla replied:
I would also do `if(this->fault)` instead of `if(this->fault == true)`.

schaubh replied:
I agree. If the fault info is coming from an optional input message, we don't need `fault` at all.  Rather, if a fault message is connected, then the fault is read in and applied.  If we temporarily don't want a fault, then the message value of `faultCapacityRatio` would just be set to 1?

juan-g-bonilla commented on src/simulation/power/simpleBattery/simpleBattery.cpp:
Hate to be this guy, but the commit message is a bit too long. As you can see in the title, the word "message" gets split because the total length is >72. It should be easy to do an interactive rebase and rename this commit :)

juan-g-bonilla commented on src/simulation/power/simpleBattery/_UnitTest/test_simpleBatteryFault.py:
```diff
@@ -0,0 +1,125 @@
+params_set_data = [(5., 5., 5., 10., 0.3),
+                    (1., 1., 5., 10., 0.3),
+                    (5, 5, 5, 10, 1e-3),
+                    (5., -5., 5., 10., 0.5)]
```
I think it'd be good to test both `faultCapacityRatio = 0` and `faultCapacityRatio = 1`, since those are the nominal limits of this parameter.

juan-g-bonilla commented on src/simulation/power/simpleBattery/simpleBattery.rst:
```diff
@@ -36,3 +36,10 @@ The next step is to attach one or more :ref:`PowerNodeUsageMsgPayload` instances
 
 For more information on how to set up and use this module, see the simple power system example :ref:`scenarioPowerDemo`.
+
+Users must configure the fault message to simulate a battery capacity fault that reduces the actual storage capacity while the ``storageCapacity`` value remains unchanged. The faulted battery capacity is determined using the ``faultCapacityRatio``, calculated as (actual capacity) / (set capacity). ::
```
Maybe reword this? Especially, I wouldn't frame it as "must" (since that implies connecting this message is mandatory), but rather "may".

> To simulate a battery capacity fault that reduces the actual storage capacity without updating the ``storageCapacity`` variable, users may connect to the ``batteryFaultInMsg`` message. The effective capacity is then ``effectiveCapacity = batteryFaultInMsg()->faultCapacityRatio * storageCapacity``.

schaubh replied:
Thanks for the feedback @juan-g-bonilla .  I'm still catching up on paper reviews for current conference, and papers for the AAS GNC ;-). I'll catch up on this review soon.

Yume27 replied:
Thank you for the reviews @juan-g-bonilla! I just updated those scripts with your comments!

juan-g-bonilla reviewed (approved):
Great PR! Left some really minor suggestions, but otherwise LGTM!

schaubh reviewed (approved):
I'm good with this PR now. I moved the commit that added the new fault message declaration to the top of the stack.  Otherwise, if someone pulls an earlier commit the code would not compile.  
Also added a minor fix to another script to address matplotlib warning.

Yume27 commented:
> I'm good with this PR now. I moved the commit that added the new fault message declaration to the top of the stack. Otherwise, if someone pulls an earlier commit the code would not compile. Also added a minor fix to another script to address matplotlib warning.
